### ISSUE TITLE
Add RecordUnmapper construction API

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#5896]
This feature extends https://github.com/jOOQ/jOOQ/issues/2520:

It should be possible to implement an "unmapper" by pairing `Field` and  `Function` as such:

```java
// R is the target record type, E is the user POJO type
interface RecordUnmapper<E, R extends Record> {
  <T> RecordUnmapper<E, R> add(Field<T> field, Function<E, T> function);
}
```

This can then be used as such:

```java
unMapper
    .add(field("COL1"), e -> e.getCol1().value)
    .add(field("COL2"), PojoClass::getCol2);
```

----
See also:
https://groups.google.com/d/msg/jooq-user/7hfDlmEwfQI/Z_F86FM0CAAJ

**Comment by steinard** [comment]
I am very interested in this functionality.

**Comment by thomasgillet** [comment]
I would be interested too. I use my own `TwoWayFieldMapper`, which is both a `RecorMapper `and `RecordUnmapper`. Could help if the unmapping part was natively supported by jOOQ.

Here a simplified version:

```java
public interface TwoWayMapper<R extends Record, E> extends RecordMapper<R, E> {
	public R unmap(E e);
}

public class TwoWayFieldMapper<R extends Record, E> implements TwoWayMapper<R, E> {
	
	private final Supplier<R> rFactory;
	private final Supplier<E> eFactory;
	private BiConsumer<R, E> r2e = null;
	private BiConsumer<E, R> e2r = null;
	
	public TwoWayFieldMapper(Supplier<R> rFactory, Supplier<E> eFactory) {
		this.rFactory = rFactory;
		this.eFactory = eFactory;
	}

	public <T> TwoWayFieldMapper<R, E> add(Field<T> field, Function<E, T> getter, BiConsumer<E, T> setter) {
		return add(
			(r, e) -> setter.accept(e, r.get(field)),
			(e, r) -> r.set(field, getter.apply(e))
		);
	}
	
	// Here there is room for various helpers:
	// - converter support (since Converter already works both ways),
	// - "embedded" classes (which already exists to some degree in DefaultRecordMapper),
	// - nested results...
	
	private TwoWayFieldMapper<R, E> add(BiConsumer<R, E> from, BiConsumer<E, R> to) {
		r2e = r2e == null ? from : r2e.andThen(from);
		e2r = e2r == null ? to : e2r.andThen(to);
		return this;
	}
	
	@Override
	public E map(R record) {
		E e = eFactory.get();
		r2e.accept(record, e);
		return e;
	}

	@Override
	public R unmap(E e) {
		R r = rFactory.get();
		e2r.accept(e, r);
		return r;
	}
}
```

**Comment by lukaseder** [comment]
Yeah, perhaps that sort of utility will be useful later on. Thanks for sharing.

**Comment by lukaseder** [comment]
https://github.com/jOOQ/jOOQ/issues/2520 is now implemented for 3.10. We can now look into implementing this.

**Comment by lukaseder** [comment]
I'm coming back to this issue time and again, and I still don't think there has ever been a clear picture regarding this topic.

10s of minor releases later, it's obvious that the `RecordMapper` "direction" (from database to client code) is optimally implemented with `MULTISET` nested collections, `ROW` nested records, and ad-hoc conversion, all starting with jOOQ 3.15.

I have a gut feeling that the inverse should be implemented in a similar way, where arbitrary nested user data structures can be transformed into tables using either `JSON_TABLE` or `XMLTABLE`, or just `UNNEST`, along with `LATERAL` and with their native unnesting capabilities. Oracle's JSON-relational duality views are also a promising topic, which jOOQ might be able to emulate to some extent:

- https://github.com/jOOQ/jOOQ/issues/14912

In any case, I don't have a clear picture at all regarding the necessity of a `RecordUnmapper` construction API, nor do I believe that configuring unmapping through SPIs will be the right approach. It's good to have this option in order to override the default for edge cases, but it won't be a strategic API in the near future.

The ideas exposed in this comment here are interesting: https://github.com/jOOQ/jOOQ/issues/5896#issuecomment-288079468

But again, I don't think this is what most people really want: A per-table based mapper/unmapper that needs to be configured globally.

Closing this as won't fix.
